src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,259 @@
+package is.hail.methods
+
+import is.hail.utils._
+import is.hail.variant._
+import is.hail.expr._
+import is.hail.keytable.KeyTable
+import is.hail.stats.{LargeNSkatModel, RegressionUtils, SkatModel}
+import is.hail.annotations.{Annotation, Querier}
+import breeze.linalg._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+
+case class SkatStat(q: Double, pval: Double)
+
+case class SkatTuple[T <: Vector[Double]](q: Double, xw: T, qtxw: DenseVector[Double])
+
+object SkatAgg {
+  val zeroValDense  = SkatAgg(0.0, new ArrayBuilder[DenseVector[Double]](), new ArrayBuilder[DenseVector[Double]]())
+  val zeroValSparse = SkatAgg(0.0, new ArrayBuilder[SparseVector[Double]](), new ArrayBuilder[DenseVector[Double]]())
+
+  def seqOp[T <: Vector[Double]](sa: SkatAgg[T], st: SkatTuple[T]): SkatAgg[T] =
+    SkatAgg[T](sa.q + st.q, sa.xws + st.xw, sa.qtxws + st.qtxw)
+
+  def combOp[T <: Vector[Double]](sa: SkatAgg[T], sa2: SkatAgg[T]): SkatAgg[T] =
+    SkatAgg[T](sa.q + sa2.q, sa.xws ++ sa2.xws.result(), sa.qtxws ++ sa2.qtxws.result())
+
+  def denseResultOp(sa: SkatAgg[DenseVector[Double]], sigmaSq: Double): SkatStat = {
+
+    val m = sa.xws.length
+    val n = sa.xws(0).size
+    val k = sa.qtxws(0).length
+
+    var xwArray = new Array[Double](m * n)
+    var qtxwArray = new Array[Double](m * k)
+    var j = 0; var i = 0
+    while(j < m * n){xwArray(j) = 0.0; j += 1}
+    while(j < m * k){qtxwArray(i) = 0.0; i += 1}
+
+    //copy in non-zeros to weighted genotype matrix array
+    i = 0
+    while (i < m) {
+      j = 0
+      while (j < n) {
+        xwArray(i * n + j) = sa.xws(i)(j)
```

Comment:
made the changes, does it matter if I make the changes to the class which handles processing the R code?